lib/SDIR/Translate/JsonEmitter.cpp

```diff
@@ -0,0 +1,215 @@
+    os.resetColor();
+    error = true;
+  }
+  symStack.push_back(SYM::BRACE);
+  indent();
+  newLine();
+  firstEntry = true;
+}
+
+void JsonEmitter::startNamedObject(StringRef name) {
+  startEntry();
+  printString(name);
+  printLiteral(": ");
+  printLiteral("{");
+  if (symStack.empty() || symStack.back() == SYM::SQUARE) {
+    // Can't have keyed values as root object or in a list
+    os.changeColor(os.RED, /*Bold=*/true);
+    printLiteral(" <<<<<<<<<<<< Started keyed object in a list");
+    os.resetColor();
+    error = true;
+  }
+  symStack.push_back(SYM::BRACE);
+  indent();
+  newLine();
+  firstEntry = true;
+}
+
+void JsonEmitter::endObject() {
+  newLine();
+  unindent();
+  tryPop(SYM::BRACE);
+  printLiteral("}");
+  firstEntry = false;
+}
+
+void JsonEmitter::startEntry() {
+  if (!firstEntry)
+    printLiteral(",");
+  firstEntry = false;
+  newLine();
+}
+
+void JsonEmitter::printKVPair(StringRef key, StringRef val, bool stringify) {
+  startEntry();
+  printString(key);
+  printLiteral(": ");
+  if (stringify)
+    printString(val);
+  else
+    printLiteral(val);
+}
+
+void JsonEmitter::printKVPair(StringRef key, int val, bool stringify) {
+  startEntry();
+  printString(key);
+  printLiteral(": ");
+  if (stringify)
+    printLiteral("\"");
+  os << val;
+  if (stringify)
+    printLiteral("\"");
+}
+
+void JsonEmitter::printKVPair(StringRef key, Attribute val, bool stringify) {
+  startEntry();
+  printString(key);
+  printLiteral(": ");
+  if (StringAttr strAttr = val.dyn_cast<StringAttr>()) {
+    strAttr.print(os);
+  } else {
+    if (stringify)
+      printLiteral("\"");
+    val.print(os);
+    if (stringify)
+      printLiteral("\"");
+  }
+}
+
+void JsonEmitter::startNamedList(StringRef name) {
+  startEntry();
+  printString(name);
+  printLiteral(": ");
+  printLiteral("[");
+  if (!symStack.empty() && symStack.back() == SYM::SQUARE) {
+    // Can't have keyed values in a list
+    os.changeColor(os.RED, /*Bold=*/true);
+    printLiteral(" <<<<<<<<<<<< Started keyed list in a list");
+    os.resetColor();
+    error = true;
+  }
+  symStack.push_back(SYM::SQUARE);
+  indent();
+  newLine();
+  firstEntry = true;
+}
+
+void JsonEmitter::endList() {
+  newLine();
+  unindent();
+  tryPop(SYM::SQUARE);
+  printLiteral("]");
+  firstEntry = false;
+}
+
+void JsonEmitter::tryPop(SYM sym) {
+  if (symStack.empty()) {
+    if (sym == SYM::BRACE)
+      printLiteral("{");
+    if (sym == SYM::SQUARE)
+      printLiteral("[");
+    os.changeColor(os.RED, /*Bold=*/true);
+    printLiteral(" <<<<<<<<<<<< Auto-inserted opening clause");
+    os.resetColor();
+    newLine();
+    indent();
+    error = true;
+  } else if (symStack.back() != sym) {
+    if (sym == SYM::BRACE)
+      printLiteral("]");
+    if (sym == SYM::SQUARE)
+      printLiteral("}");
+    os.changeColor(os.RED, /*Bold=*/true);
+    printLiteral(" <<<<<<<<<<<< Auto-inserted closing clause");
```

Comment:
If there is a mismatch of opening/closing braces/brackets then there is a bug in the translator. This should never happen regardless of the input MLIR. This line clearly shows the issue in the output (the color and printout is hard to miss when checking the output) and therefore makes debugging the translator easier. It would be much harder to track down the source of the bug if we would simply produce invalid JSON. E.g. if we know that the translator forgot to close the `edges` array, then we can check the function(s) that produce said array.

Reply:
ok